[itsvari] iP

itsvari opened the pull request:
# SONGBIRD™ AI Assistant Mark IV 🤖 
### by VariTech Heavy Industries, ©️ 3025. All rights reserved and enforced by nanite swarm.

> "The only AI assistant with full military-grade neural encryption and integrated thought synchronization capabilities." - Dr. Elena Volkov, Chief of Cybernetic & Autonomic Research

Welcome to the official repository of SONGBIRD™, your personal AI productivity enhancement system. This state-of-the-art digital consciousness has been carefully engineered by VariTech's elite team of neural architects to serve your every ~~surveillance~~ organizational need.

## Features
* **Military-grade** task management protocols
* _Quantum-encrypted_ storage capabilities
* Advanced ~~mind control~~ reminder systems
* Multi-dimensional datetime parsing
* Neural-adaptive user interface

## System Requirements
1. Minimum quantum processing unit: Mark VII or higher
1. Neural interface port (~~mandatory~~ optional)
1. Minimum 1TB neural bandwidth
1. VariTech-approved consciousness stabilizer

## Quick Start
```java
public static void main(String[] args) {
    Songbird songbird = new Songbird();
    songbird.init();  // Initializes neural link
}
```

## Implementation Status
* [x] Basic consciousness implementation
* [x] Task management protocols
* [x] Storage encryption
* [ ] Full neural integration
* [ ] Quantum teleportation capabilities
* [ ] Time manipulation features

## Safety Protocols
* Never look directly into SONGBIRD's digital consciousness. Any damage to one's sanity, loss of will, or death will not be covered under the Limited Warranty.
* Maintain regular neural backups
* Keep your [VariTech Mental Firewall™](https://itsvari.com) active at all times
* Use command `bye` to safely disconnect neural interface

## Legal Disclaimer
By using SONGBIRD™, you agree to allow VariTech Heavy Industries access to your:
* **Thoughts**
* _Dreams_
* ~~Hopes~~
* Memories
* ~~Soul~~

For technical support, please direct your thoughts to our neural support network or consult your assigned VariTech consciousness technician.

---
_VariTech Heavy Industries - Resistance Is Obsolete_

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

itsvari committed:
Update .gitignore

itsvari committed:
Complete all Level-0 objectives

itsvari committed:
Extract UI-handling code to new class

itsvari committed:
Echo user commands as per Level-1 spec

itsvari committed:
Create abstract Command class and related enum

itsvari committed:
Create classes to handle Tasks

itsvari committed:
Create multiple Command types with associated behaviour

itsvari committed:
Create Parser to parse commands

itsvari committed:
Update UI for more flair

itsvari committed:
Update main class to handle new logic

itsvari committed:
Linting: fixed minor issues

itsvari committed:
Rename packages

itsvari committed:
Remove Duke.java

itsvari committed:
Add error handling to Ui.java

itsvari committed:
Add string parsing to CommandType enum

itsvari committed:
Add custom exceptions

itsvari committed:
Add new logic to handle marking tasks as done or not done

itsvari committed:
Remove unnecessary toString() calls

itsvari committed:
Add QoL improvements to TaskList

itsvari committed:
Add support for different Task types and error handling

itsvari committed:
Update text UI testing suite

itsvari committed:
Add delete command and additional error handling

itsvari committed:
Add delete command to test suite

itsvari committed:
Delete unused exception

itsvari committed:
Fix javadoc errors and add missing javadocs

itsvari committed:
Simplify task count message handling

itsvari committed:
Merge branch 'master' into branch-Level-7

itsvari committed:
Add new exception class for storage exceptions

itsvari committed:
Fix bye command to work with new main loop

itsvari committed:
Fix minor formatting errors

itsvari committed:
Add persistent storage support for tasks

itsvari committed:
Merge branch 'branch-Level-7'

itsvari committed:
Change how indexing is handled

itsvari committed:
Change error message styling

itsvari committed:
Update .gitignore

itsvari committed:
Modify tasks to use proper date datatypes

itsvari committed:
Modify greeting UI

itsvari committed:
Merge branch 'branch-Level-8'

itsvari committed:
Merge branch 'add-gradle-support'

itsvari committed:
Update Gradle dependencies

itsvari committed:
Update getAll to package-private for use in tests

itsvari committed:
Add Mockito to dependencies

itsvari committed:
Add first round of JUnit tests

itsvari committed:
Merge branch 'branch-A-JUnit'

itsvari committed:
Update Gradle .jar build instructions

itsvari committed:
Merge branch 'branch-A-Jar'

itsvari committed:
Add checkstyle config files

itsvari committed:
Add checkstyle configs to build.gradle

itsvari committed:
Merge branch 'branch-A-Checkstyle'

itsvari committed:
Fix code style violations

itsvari committed:
Add additional javadocs

itsvari committed:
Add find command

itsvari committed:
Add unit tests for new find command

itsvari committed:
Merge branch 'branch-A-CodingStandard'

itsvari committed:
Merge branch 'branch-A-JavaDoc'

itsvari committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/test/java/songbird/task/TaskListTest.java

itsvari committed:
Fix style violations

itsvari committed:
Add GitHub CI support

itsvari committed:
Merge branch 'branch-A-CI'

itsvari committed:
Fix text encoding issues on Windows

itsvari committed:
Add deletion delay to ensure resource cleanup

itsvari committed:
Merge branch 'bugfix-windows-failing-tests'

itsvari committed:
Add devcontainer support

itsvari committed:
Merge branch 'devops-devcontainer'

itsvari committed:
Add automatic release build action

itsvari committed:
Update README

itsvari committed:
Add dependabot support

itsvari committed:
Revise gradle CI pipeline

itsvari committed:
Fix issues regarding caching

itsvari committed:
Merge branch 'devops-misc-ci'

itsvari committed:
Add AtlantaFX to dependencies

itsvari committed:
Merge branch 'master' into branch-A-JavaDoc

itsvari committed:
Update Javadocs to comply with code style

itsvari committed:
Merge branch 'branch-A-JavaDoc'

samuelneo commented on src/main/java/songbird/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,188 @@
+ * @author Ashe Low
+ * @version CS2103T AY24/25 Semester 2
+ */
+public class DateTimeParser {
```
I like your implementation of the date/time parser, it's very robust and lenient towards various input formats

itsvari replied:
haha thanks! it definitely doesn't cover every format i want it to, i'll probably switch to using something like [hawking](https://github.com/zoho/hawking) instead. after all, someone else has done all the legwork, might as well use that ;)

samuelneo commented on src/main/java/songbird/task/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+     *
+     * @return A copy of the task list.
+     */
+    List<Task> getAll() {
```
is this method intentionally package-private? I think (unless Ctrl-F is failing me) it's only accessed once and within this class, so maybe it could be private?

itsvari replied:
it's used in `TaskListTest::testGetAllEncapsulation`, so that's why it's package-private!

```java
/**
 * Tests the encapsulation of the getAll method.
 * Verifies that modifications to the returned list don't affect the original list.
 */
@Test
public void testGetAllEncapsulation() {
    // Arrange
    Task task = new ToDoTask("Read a book");
    taskList.addTask(task);


    // Act
    List<Task> allTasks = taskList.getAll();


    // Attempt to modify the returned list
    allTasks.add(new ToDoTask("New Task"));


    // Assert
    assertEquals(1, taskList.getSize(),
            "Modifying the returned list should not affect the original TaskList.");
}
```

samuelneo replied:
ahh I see

samuelneo commented on src/main/java/songbird/parser/Parser.java:
```diff
@@ -0,0 +1,185 @@
+                String eventStartString;
+                String eventEndString;
+
+                // must be in format: <DESCRIPTION> /from <START> /to <END>
+                if (!taskParameters.contains(" /from ") || !taskParameters.contains(" /to ")) {
+                    throw new SongbirdMalformedCommandException("The event task must have a start and end time, "
+                            + "formatted as '/from YYYY-MM-DD HH:MM /to YYYY-MM-DD HH:MM'.");
+                }
+
+                // Split the taskParameters string
+                String[] fromParts = taskParameters.split(" /from ", 2);
+                String[] toParts;
+
+                // extract the description, start, and end times, no matter the order
+                if (fromParts[1].contains(" /to ")) {
+                    // if /from comes before /to
+                    eventDescription = fromParts[0].trim();
+                    toParts = fromParts[1].split(" /to ", 2);
+                    eventStartString = toParts[0].trim();
+                } else {
+                    // if /to comes before /from
+                    eventStartString = fromParts[1].trim();
+                    toParts = fromParts[0].split(" /to ", 2);
+                    eventDescription = toParts[0].trim();
+                }
+                eventEndString = toParts[1].trim();
```
personally I found this section easier to parse using regex (though what you've done is definitely fine too, and it's nice that you've added support for having `/to` before `/from`)

itsvari replied:
regex would work too! but the regex required to handle fuzzy from/to matching is.... yuck:
`(.+?)\s+/(?:(?:from\s+(.+?)\s+/to)|(?:to\s+(.+?)\s+/from))\s+(.+)`

overall i felt that doing string splitting was a bit easier to comprehend, and also provides more extensibility if we want to detect alternate keywords in the future!

samuelneo replied:
yeahh the regex gets messier if you're supporting fuzzy from/to matching haha

samuelneo commented on src/test/java/songbird/storage/StorageTest.java:
```diff
@@ -0,0 +1,293 @@
+                // Ensure file handle is released by forcing GC and waiting a bit
+                System.gc();
+                Thread.sleep(100);
+                boolean deleted = file.delete();
```
coding standard: should be named to sound like a boolean (something like `isDeleted`)
this appears in a couple other places too (lines 81 and 246 of `StorageTest.java`)

itsvari replied:
good catch lol, i wrote this code at ~6am at the end of a 7-hour sprint while extremely sleep-deprived haha ^-^

samuelneo reviewed (commented):
very well-written and well-documented code with thorough tests; I was struggling to come up with comments

itsvari committed:
Fix boolean naming to fit code conventions

zihung20 commented on src/main/java/songbird/command/Command.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Executes the command.
+     */
+    public abstract void execute();
```
I like how you designed your Command class. Each subclass of Command implements `execute()`, ensuring they have their own behavior.

wzhua02 commented on src/main/java/songbird/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,188 @@
+            try {
+                return LocalDateTime.parse(input, DateTimeFormatter.ofPattern(pattern));
+            } catch (DateTimeParseException e) {
+                // ignore and try next pattern
```
Perhaps include a continue here? In case you add something after try catch? But not wrong here.

itsvari replied:
Continue is not needed here, even if we were to place something after the try/catch block; execution would just flow on further down.

wzhua02 commented on src/main/java/songbird/task/Task.java:
```diff
@@ -0,0 +1,67 @@
+ * @author Ashe Low
+ * @version CS2103T AY24/25 Semester 2
+ */
+public abstract class Task implements Serializable {
```
Was wondering why do you implement Serializable and what's its purpose?

itsvari replied:
You need to implement Serializable to use Java's built in SerDes operations, in order to write it to a file

wzhua02 commented on src/main/java/songbird/command/CommandType.java:
```diff
@@ -0,0 +1,46 @@
+ * @author Ashe Low
+ * @version CS2103T AY24/25 Semester 2
+ */
+public enum CommandType {
```
Wondering why do you specify a command type class when each command already has a class?

itsvari replied:
I find it much cleaner to use an enum combined with a switch, rather than constantly doing `InstanceOf` checks and chained if-else statements, when doing command processing.

IIRC, Java 17 does have support for type pattern matching, but only as a preview. I just went with an enum instead.

wzhua02 commented on src/main/java/songbird/parser/Parser.java:
```diff
@@ -0,0 +1,185 @@
+     * @param input The user input to be parsed.
+     * @return The command object that corresponds to the user input.
+     */
+    public Command parse(String input) throws SongbirdException {
```
I like the switch cases done in the parser.

wzhua02 commented on src/main/java/songbird/Songbird.java:
```diff
@@ -0,0 +1,102 @@
+    public Songbird() {
+        try {
+            this.ui = new Ui();
+            Storage storage = new Storage("./data/tasklistDB");
```
Perhaps can move the filepath into the storage class?

itsvari replied:
It's for extensibility, to allow multiple Storage instances to be initialised. Better not to hardcode magic/hidden filepaths into the Storage class itself

wzhua02 commented on src/test/java/songbird/task/DeadlineTaskTest.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Test
+    public void testConstructor() {
```
Perhaps change the test method name format to "unitBeingTested_descriptionOfTestInputs_expectedOutcome". Also could include some javadocs.

itsvari replied:
yup, good catch

wzhua02 reviewed (commented):
The code is really well written, good job.

zihung20 commented on src/main/java/songbird/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,188 @@
+        "yyyy/MM/dd hh a",
+        "yyyy-MM-dd h a",
+        "yyyy/MM/dd h a",
+        "dd-MM-yyyy HH:mm",
+        "dd/MM/yyyy HH:mm",
+        "dd-MM-yyyy H:mm",
+        "dd/MM/yyyy H:mm",
+        "dd-MM-yyyy HHmm",
+        "dd/MM/yyyy HHmm",
+        "dd-MM-yyyy HH",
+        "dd/MM/yyyy HH",
+        "dd-MM-yyyy H",
+        "dd/MM/yyyy H",
+        "dd-MM-yyyy hh:mm a",
+        "dd/MM/yyyy hh:mm a",
+        "dd-MM-yyyy hh:mma",
+        "dd/MM/yyyy hh:mma",
+        "dd-MM-yyyy h:mma",
+        "dd/MM/yyyy h:mma",
+        "dd-MM-yyyy h:mm a",
+        "dd/MM/yyyy h:mm a",
+        "dd-MM-yyyy hh a",
+        "dd/MM/yyyy hh a",
+        "dd-MM-yyyy h a",
+        "dd/MM/yyyy h a",
+        "MM/dd/yyyy HH:mm",
+        "MM-dd-yyyy HH:mm",
```
Wouldn't there be cases where months are misinterpreted? For example, if a user enters 11-01-2025 12:34 intending it as November 1st, but due to the sequence, it gets interpreted as January 11th instead? Or this is your intentional implementation?

itsvari replied:
yep, that's a very good catch! it's hard to avoid ambiguity without providing a configuration option.

though, it could be customized so that it prioritizes the system locale's way of writing dates... 🤔

but yeah, i18n and l10n is *hard*.

zihung20 reviewed (commented):
Overall, I found this PR flawless. It’s hard to find any issues since exceptions are well handled, the files are well-structured, and the test cases comprehensively cover the class behavior. **Very good work!** 😊

itsvari committed:
Add JavaFX to Gradle dependencies

itsvari committed:
Remove UITest

itsvari committed:
Add user and agent assets

itsvari committed:
Add MVP UI

itsvari committed:
Fix tests

itsvari committed:
Prevent Songbird object reassignment

itsvari committed:
Merge branch 'branch-Level-10'

itsvari committed:
Improve unit test names and Javadocs

itsvari committed:
Refactor CommandType enum

itsvari committed:
Refine variable names

itsvari committed:
Refactor Parser to adhere to SLAP

The Parser class is handling the parsing, preparation, and execution of
every possible command type within a single monolithic method
Parser::parse. This simultaneously breaks the long method and SLAP
principles of code quality laid out in the code quality standards.

Breaking out the handling of each command into its own method thus
significantly improves code quality, by adhering to the SLAP principle,
and makes each method significantly shorter. This boosts readability and
maintainability.

itsvari committed:
Add assertions to DialogBox

itsvari committed:
Merge pull request #7 from itsvari/branch-A-Assertions

Add assertions to DialogBox

itsvari committed:
Merge branch 'master' of github.com-itsvari:itsvari/ip

itsvari committed:
Merge branch 'master' into branch-A-CodeQuality

itsvari committed:
Merge pull request #8 from itsvari/branch-A-CodeQuality

Initial pass at improving code quality.

This commit introduces a significant refactor to the Parser class.
The Parser class is handling the parsing, preparation, and execution of
every possible command type within a single monolithic method
Parser::parse. This simultaneously breaks the long method and SLAP
principles of code quality laid out in the code quality standards.

Breaking out the handling of each command into its own method thus
significantly improves code quality, by adhering to the SLAP principle,
and makes each method significantly shorter. This boosts readability and
maintainability.

Other miscellaneous fixes for code quality include improving variable
names in TaskList and improving the handling of string descriptions in the
CommandType enum.

itsvari committed:
Merge branch 'master' into branch-A-CI

itsvari committed:
Fix CI for auto-triggered releases

For some reason, this CI only works when automatically triggered, but not in other contexts. This additional checkout action should hopefully fix the failing CI runs.

itsvari committed:
Merge branch 'branch-A-CI'

itsvari committed:
Add Hawking for natural language date processing

itsvari committed:
Use Hawking for natural language time processing

itsvari committed:
Update to use LocalDateTime where possible

itsvari committed:
Fix code style issues

itsvari committed:
Merge branch 'branch-C-NaturalDates'

We currently are handling date and time parsing by using an extremely
long list of possible date, time, and date + time formats, and trying
those formatters one-by-one using Java's built-in date and time parsing
commands.

However, this is a relatively brittle implementation, and does
not cover the wide-ranging set of possible date and time formats. Trying
to use this method to capture the multiple permutations of possible date
and time formats is not feasible if we want to enable natural language
date processing.

Let's change the date and time parser to use the Hawking natural
language date/time parser library, so that we can use natural language
processsing to parse dates. Since natural language processing is
probabilistic, we will still keep a more rigid command
syntax. This reduces the chance of a failure to interpret the
user-specified time, and should make for a better UX.

itsvari committed:
Merge branch 'master' into branch-A-CI

itsvari committed:
Update dependencies

itsvari committed:
Merge branch 'branch-A-CI'

itsvari committed:
Add Ai.md report

itsvari committed:
Merge branch 'branch-A-AiAssisted'

itsvari committed:
Fix typos

itsvari committed:
Fix app launch regressions

itsvari committed:
Deprecate text UI testing

itsvari committed:
Add first pass of documentation

itsvari committed:
Modify welcome message

itsvari committed:
Add response buffer to Ui

Currently, if there are storage errors, or if no existing storage file is found, we immediately try to send a status message indicating as such.

However, since no MainWindow has been created at that point in the program's lifecycle, this leads to a NullPointerException and causes the app to crash at launch.

Let's buffer responses if there is no mainWindow instance present, and add a check to respond() to print out the messages if there are buffered responses.

Since we always call a respond() a second time to show the welcome banner, this will not cause a delay in the status message being sent.

itsvari committed:
Fix app crashing with NullPointerException

itsvari committed:
Add window title

itsvari committed:
Merge branch 'branch-A-UserGuide'

itsvari committed:
Adjust styling of messages and icons

itsvari committed:
Merge branch 'branch-A-BetterGui'

itsvari committed:
Adjust UI sizing

itsvari committed:
Add product screenshot

itsvari committed:
Add AI disclaimer to README

itsvari committed:
Change release version number
